utilities/pspp-convert.1
.\" -*- nroff -*-
.\" Copyright (C) 2020  Free Software Foundation

.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.

.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.

.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.de IQ
.  br
.  ns
.  IP "\\$1"
..
.TH pspp\-convert 1 "October 2013" "PSPP" "PSPP Manual"
.
.SH NAME
pspp\-convert \- convert SPSS data files to other formats
.
.SH SYNOPSIS
\fBpspp\-convert\fR [\fIoptions\fR] \fIinput\fR \fIoutput\fR
.br
\fBpspp\-convert \-\-help\fR | \fB\-h\fR
.br
\fBpspp\-convert \-\-version\fR | \fB\-v\fR
.
.SH DESCRIPTION
The \fBpspp\-convert\fR program reads \fIinput\fR, which may be an
SPSS system file, an SPSS/PC+ system file, an SPSS portable file,
or an encrypted SPSS syntax file,
and writes it to \fIoutput\fR, performing format conversion as
necessary.
.PP
The format of \fIinput\fR is automatically detected, when possible.
The character encoding of old SPSS system files cannot always be
guessed correctly, and SPSS/PC+ system files do not include any
indication of their encoding.  Use \fB\-e \fIencoding\fR to specify
the encoding in this case.
.PP
By default, the intended format for \fIoutput\fR is inferred from its
extension:
.
.IP \fBcsv\fR
.IQ \fBtxt\fR
Comma-separated value.  Each value is formatted according to its
variable's print format.  The first line in the file contains variable
names.
.
.IP \fBsav\fR
.IQ \fBsys\fR
SPSS system file.
.
.IP \fBpor\fR
SPSS portable file.
.
.IP \fBsps\fR
SPSS syntax file.  (Only encrypted syntax files may be converted to
this format.)
.
.PP
Use \fB\-O \fIextension\fR to override the inferred format or to
specify the format for unrecognized extensions.
.
.PP
\fBpspp\-convert\fR can convert most input formats to most output
formats.  Encrypted SPSS file formats are exceptions: if the input
file is in an encrypted format, then the output file will be the same
format (decrypted).  Options for the output format are ignored in this
case.
.
.SH "OPTIONS"
.SS "General Options"
.
.IP "\fB\-O format\fR"
.IQ "\fB\-\-output\-format=\fIformat\fR"
Specifies the desired output format.  \fIformat\fR must be one of the
extensions listed above, e.g. \fB\-O csv\fR requests comma-separated
value output.
.
.IP "\fB\-c \fImaxcases\fR"
.IQ "\fB\-\-cases=\fImaxcases\fR"
By default, all cases are copied from \fIinput\fR to \fIoutput\fR.
Specifying this option to limit the number of cases written to
\fIoutput\fR to \fImaxcases\fR.
.
.IP "\fB\-e \fIcharset\fR"
.IQ "\fB\-\-encoding=\fIcharset\fR"
Overrides the encoding in which character strings in \fIinput\fR are
interpreted.  This option is necessary because old SPSS system files
do not self-identify their encoding.
.
.IP "\fB\-k \fIvar\fR..."
.IQ "\fB\-\-keep=\fIvar\fR..."
Drops all variables except those listed as \fIvar\fR, and reorders the
remaining variables into the specified order.
.
.IP "\fB\-d \fIvar\fR..."
.IQ "\fB\-\-drop=\fIvar\fR..."
Drops each \fIvar\fR listed from the output.
.
.SS "CSV Output Options"
.PP
These options affect only output to \fB.csv\fR and \fB.txt\fR files.
.IP "\fB\-\-labels\fR"
By default, \fBpspp\-convert\fR writes variables' values to the output.
With this option, \fBpspp\-convert\fR writes value labels.
.IP "\fB\-\-no\-var\-names\fR"
By default, \fRpspp\-convert\fR\fR writes the variable names as the
first line of output.  With this option, \fBpspp\-convert\fR omits
this line.
.IP "\fB\-\-recode\fR"
By default, \fBpspp\-convert\fR writes user-missing values as their
regular values.  With this option, \fBpspp\-convert\fR recodes them to
system-missing values (which are written as a single space).

.IP "\fB\-\-print\-formats\fR"
By default, \fBpspp\-convert\fR writes numeric variables as plain
numbers.  This option makes \fBpspp\-convert\fR honor variables'
print formats.

.IP "\fB\-\-decimal=\fIdecimal\fR"
This option sets the character used as a decimal point in output.  The
default is a period (\fB.\fR).

.IP "\fB\-\-delimiter=\fIdelimiter\fR"
This option sets the character used to separate fields in output.  The
default is a comma (\fB,\fR), unless the decimal point is a comma, in
which case a semicolon (\fB;\fR) is used.

.IP "\fB\-\-qualifier=\fIqualifier\fR"
The option sets the character used to quote fields that contain the
delimiter.  The default is a double quote (\fB\(dq\fR).
.
.SS "Password Options"
When the input file is encrypted, \fBpspp\-convert\fR needs to obtain
a password to decrypt it.  To do so, the user may specify the password
with \fB\-p\fR (or \fB\-\-password\fR), or the name of a file containing a
list of possible passwords with \fB\-\-password\-list\fR, or an
alphabet of possible passwords to try along with a maximum length with
\fB\-a\fR (or \fB\-\-password\-alphabet\fR) and \fB\-l\fR (or
\-\-password\-length\fR).  If none of these options is used,
\fBpspp\-convert\fR prompts for the password.  The password options
are:
.
.IP "\fB\-p \fIpassword\fR"
.IQ "\fB\-\-password=\fIpassword\fR"
Specifies the password to use to decrypt the input file.
.
.IP
On multiuser systems, this option may not be safe because other users
may be able to see the password in process listings.
.
.IP "\fB\-a \fIalphabet \fB\-l \fImax-length\fR"
.IQ "\fB\-\-password-alphabet=\fIalphabet\ \fB\-\-password-length=\fImax-length\fR"
These options are an alternative to \fB\-p\fR or \fB\-\-password\fR.
They direct \fBpspp\-convert\fR to search for the correct password
from the set of all passwords of symbols from \fIalphabet\fR (which
may contain character ranges specified with \fB-\fR) and no more than
\fImax-length\fR symbols long.  For example, \fB\-a a-z \-l 5\fR
checks all possible lowercase alphabetic passwords no more than 5
characters long.
.IP
When these options are used, \fB\-p\fR may additionally specify a
starting point for the search.
.
.IP "\fB\-\-password\-list=\fIfile\fR"
Specifies a file to read containing a list of passwords to try, one
per line.  If \fIfile\fR is \fB\-\fR, reads from stdin.
.
.SS "Other Options"
.IP "\fB\-h\fR"
.IQ "\fB\-\-help\fR"
Prints a usage message on stdout and exits.
.
.IP "\fB\-v\fR"
.IQ "\fB\-\-version\fR"
Prints version information on stdout and exits.
.
.SH "AUTHORS"
Ben Pfaff.
.
.SH "SEE ALSO"
.
.BR pspp\-output (1),
.BR pspp (1),
.BR psppire (1).


NEWS
PSPP NEWS -- history of user-visible changes.
Copyright (C) 1996-2000, 2008-2016, 2017, 2019, 2020 Free Software Foundation, Inc.
See the end for copying conditions.

Please send PSPP bug reports to bug-gnu-pspp@gnu.org.

Changes from 1.4.1 to 1.5.2:

 * An error in the displayed signficance of oneway anova
   contrasts tests has been corrected.

 * Added Drag-N-Drop in output view.

 * The Explore GUI dialog supports the "Plots" subdialog. Boxplots, Q-Q Plots
   and Spreadlevel plots are now also available via the GUI.

 * The graphical user interface for importing spreadsheets has been improved.
   The new interface provides the user with a preview of the data to be imported
   and interactive methods to select the desired ranges.

 * The user manual, in its Info and HTML versions, now includes
   graphical output examples and screenshots.

 * Output improvements:

   - New drivers for output to TeX source files and to PNG files.

   - Table output styles may now be set with the new option
     --table-look and the new SET TLOOK command.

   - New driver option "trim" to remove empty space from PDF,
     PostScript, SVG, and PNG output files.

   - The PDF output driver now adds an outline to allow PDF viewers to display
     as a "table of contents" for the file.

   - The HTML output driver has a new option "bare".

 * New features in pspp-output:

   - New --table-look and --nth-commands options.

   - New get-table-look and convert-table-look commands.

Changes from 1.4.0 to 1.4.1:

 * Bug fixes.

 * Translation updates.

 * HTML output is now HTML5 instead of HTML4.0 Transitional.

 * The HOST command has been reimplemented.  It now supports TIMELIMIT.

Changes from 1.2.0 to 1.4.0:

 * The file pspp-mode.el (the pspp-mode for Emacs) is no longer distributed
   with PSPP.  Instead, it can be installed directly in emacs and is available
   from https://elpa.gnu.org/packages/pspp-mode.html

 * The Find dialog box, when searching for numeric values, will match only
   to the precision of the variable's print format.  This avoids behaviour
   which is surprising to some users.

 * PSPP now supports the SPSS viewer (.spv) format that SPSS 16 and later
   use to save the contents of its output editor:

   - PSPP and PSPPIRE can write output to .spv files.

   - The new utility pspp-output can convert .spv files to other formats.

   - The pspp-convert utility can now decrypt encrypted .spv files.  The
     encrypted viewer file format is unacceptably insecure, so to
     discourage its use PSPP and PSPPIRE do not directly read or write
     this format.

 * A bug where the Data|Select Cases|Random Sample menu would generate
   invalid syntax has been fixed.

 * FREQUENCIES will emit a warning if one or more weight values are missing.

 * Plain text output is no longer divided into pages, since it is now
   rarely printed on paper.

 * pspp-convert improvements:

   - New "-a", "-l", "--password-list" options to search for an
     encrypted file's password.

   - New "--keep" and "--drop" options to output only selected variables.

   - New "--recode", "--no-var-name", "--labels", "--print-formats",
     "--decimal", "--delimiter", and "--qualifier" options to control
     CSV output.

 * Improvements to SAVE DATA COLLECTION support for MDD files.

 * Bug fix for CVE-2018-20230.

 * The /RENAME subcommand in SAVE TRANSLATE et al. has been extended
   to support the syntax also accepted by other software.

 * The EXAMINE command will now perform the Shapiro-Wilk test when
   one or more plots are requested.

 * The REGRESSION command now supports the /STATISTICS=TOL option which
   outputs tolerance and variance inflation factor metrics for the data.

 * The QUICK CLUSTER command now supports the /SAVE option which can
   be used to save the cases' cluster membership and/or their distance
   from the cluster centre to the active file.

 * The MEANS command has been re-implemented.

 * The GUI now has a sub-dialog for post-hoc tests of the oneway
   anova test.

 * Gtk+3.22.0 or later must now be used when building.

 * GRAPH: Categorical barcharts displaying percentages now show the
   sub-category as a percentage of the primary category instead of the
   total.

 * The Syntax Window now has a Find function to perform simple text
   searching.

 * AUTORECODE:

   - The PRINT subcommand is now implemented.

   - Many bug fixes.

Changes from 1.0.1 to 1.2.0:

 * New experimental command SAVE DATA COLLECTION to save MDD files.

 * MTIME and YMDHMS variable formats now supported.

 * Build changes:

   - zlib is now a required dependency.  (Previously it was optional.)

   - Gtk+3.18.0 or later must now be used when building.

   - The code implementing the sheet rendering has been removed.  Instead we
     use a third party library: spread-sheet-widget.

 * The "NOMISSING" keyword when generating syntax for piecharts in the
   frequencies command was omitted.  This has been corrected.

Changes from 1.0.0 to 1.0.1:

 * Bug fixes, including fixes for CVE-2017-12958, CVE-2017-12959,
   CVE-2017-12960, and CVE-2017-12961.

Changes from 0.11.0 to 1.0.0:

 * Translation updates.

Changes from 0.10.2 to 0.11.0:

 * The REGRESSION command now has a /ORIGIN subcommand to perform
   regression through the origin.

 * The FACTOR command can now analyse matrix files prepared with MATRIX DATA.

 * The FACTOR command can now print the anti-image matrices.

 * The MATRIX DATA command has been added.

 * Some inappropriate properties in selection dialogs have been corrected.

 * A bug which could cause the HTML driver to go into a tight loop
   has been fixed.

 * An error in the FREQUENCIES procedure, where the word "Mean" was
   printed when "Variance" was appropriate has been fixed.

 * The ncurses library is no longer required or used.

 * A bug where the Mann-Whitney test would give misleading results
   if run on multiple variables and MISSING=ANALAYSIS was specified
   has been fixed.


utilities/pspp-convert.c
/* PSPP - a program for statistical analysis.
   Copyright (C) 2013, 2014, 2015, 2016 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>. */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <stdlib.h>
#include <unistd.h>

#include "data/any-reader.h"
#include "data/casereader.h"
#include "data/casewriter.h"
#include "data/csv-file-writer.h"
#include "data/dictionary.h"
#include "data/encrypted-file.h"
#include "data/file-name.h"
#include "data/por-file-writer.h"
#include "data/settings.h"
#include "data/sys-file-writer.h"
#include "data/file-handle-def.h"
#include "language/command.h"
#include "language/lexer/lexer.h"
#include "language/lexer/variable-parser.h"
#include "libpspp/assertion.h"
#include "libpspp/cast.h"
#include "libpspp/i18n.h"

#include "gl/error.h"
#include "gl/getpass.h"
#include "gl/localcharset.h"
#include "gl/progname.h"
#include "gl/version-etc.h"

#include "gettext.h"
#define _(msgid) gettext (msgid)

static void usage (void);

static bool decrypt_file (struct encrypted_file *enc,
                          const struct file_handle *input_filename,
                          const struct file_handle *output_filename,
                          const char *password,
                          const char *alphabet, int max_length,
                          const char *password_list);

static void
parse_character_option (const char *arg, const char *option_name, char *out)
{
  if (strlen (arg) != 1)
    {
      /* XXX support multibyte characters */
      error (1, 0, _("%s argument must be a single character"), option_name);
    }
  *out = arg[0];
}

static bool
parse_variables_option (const char *arg, struct dictionary *dict,
                        struct variable ***vars, size_t *n_vars)
{
  struct lexer *lexer = lex_create ();
  lex_append (lexer, lex_reader_for_string (arg, locale_charset ()));
  lex_get (lexer);

  bool ok = parse_variables (lexer, dict, vars, n_vars, 0);
  if (ok && (lex_token (lexer) != T_STOP && lex_token (lexer) != T_ENDCMD))
    {
      lex_error (lexer, _("expecting variable name"));
      ok = false;
    }

  lex_destroy (lexer);
  if (!ok)
    {
      free (*vars);
      *vars = NULL;
      *n_vars = 0;
    }
  return ok;
}

int
main (int argc, char *argv[])
{
  const char *input_filename;
  const char *output_filename;

  long long int max_cases = LLONG_MAX;
  const char *keep = NULL;
  const char *drop = NULL;
  struct dictionary *dict = NULL;
  struct casereader *reader = NULL;
  struct file_handle *input_fh = NULL;
  const char *encoding = NULL;
  struct encrypted_file *enc;

  const char *output_format = NULL;
  struct file_handle *output_fh = NULL;
  struct casewriter *writer;
  const char *password = NULL;
  struct string alphabet = DS_EMPTY_INITIALIZER;
  const char *password_list = NULL;
  int length = 0;

  struct csv_writer_options csv_opts = {
    .include_var_names = true,
    .decimal = '.',
    .delimiter = 0,             /* The default will be set later. */
    .qualifier = '"',
  };

  long long int i;

  set_program_name (argv[0]);
  i18n_init ();
  fh_init ();
  settings_init ();

  for (;;)
    {
      enum
        {
          OPT_PASSWORD_LIST = UCHAR_MAX + 1,
          OPT_RECODE,
          OPT_NO_VAR_NAMES,
          OPT_LABELS,
          OPT_PRINT_FORMATS,
          OPT_DECIMAL,
          OPT_DELIMITER,
          OPT_QUALIFIER,
        };
      static const struct option long_options[] =
        {
          { "cases", required_argument, NULL, 'c' },
          { "keep", required_argument, NULL, 'k' },
          { "drop", required_argument, NULL, 'd' },
          { "encoding", required_argument, NULL, 'e' },

          { "recode", no_argument, NULL, OPT_RECODE },
          { "no-var-names", no_argument, NULL, OPT_NO_VAR_NAMES },
          { "labels", no_argument, NULL, OPT_LABELS },
          { "print-formats", no_argument, NULL, OPT_PRINT_FORMATS },
          { "decimal", required_argument, NULL, OPT_DECIMAL },
          { "delimiter", required_argument, NULL, OPT_DELIMITER },
          { "qualifier", required_argument, NULL, OPT_QUALIFIER },

          { "password", required_argument, NULL, 'p' },
          { "password-alphabet", required_argument, NULL, 'a' },
          { "password-length", required_argument, NULL, 'l' },
          { "password-list", required_argument, NULL, OPT_PASSWORD_LIST },

          { "output-format", required_argument, NULL, 'O' },

          { "help",    no_argument,       NULL, 'h' },
          { "version", no_argument,       NULL, 'v' },
          { NULL,      0,                 NULL, 0 },
        };

      int c;

      c = getopt_long (argc, argv, "c:k:d:e:p:a:l:O:hv", long_options, NULL);
      if (c == -1)
        break;

      switch (c)
        {
        case 'c':
          max_cases = strtoull (optarg, NULL, 0);
          break;

        case 'k':
          keep = optarg;
          break;

        case 'd':
          drop = optarg;
          break;

        case 'e':
          encoding = optarg;
          break;

        case 'p':
          password = optarg;
          break;

        case 'l':
          length = atoi (optarg);
          break;

        case OPT_PASSWORD_LIST:
          password_list = optarg;
          break;

        case OPT_RECODE:
          csv_opts.recode_user_missing = true;
          break;

        case OPT_NO_VAR_NAMES:
          csv_opts.include_var_names = false;
          break;

        case OPT_LABELS:
          csv_opts.use_value_labels = true;
          break;

        case OPT_DECIMAL:
          parse_character_option (optarg, "--decimal", &csv_opts.decimal);
          break;

        case OPT_DELIMITER:
          parse_character_option (optarg, "--delimiter", &csv_opts.delimiter);
          break;

        case OPT_QUALIFIER:
          parse_character_option (optarg, "--qualifier", &csv_opts.qualifier);
          break;

        case 'a':
          for (const char *p = optarg; *p;)
            if (p[1] == '-' && p[2] > p[0])
              {
                for (int ch = p[0]; ch <= p[2]; ch++)
                  ds_put_byte (&alphabet, ch);
                p += 3;
              }
            else
              ds_put_byte (&alphabet, *p++);
          break;

        case 'O':
          output_format = optarg;
          break;

        case 'v':
          version_etc (stdout, "pspp-convert", PACKAGE_NAME, PACKAGE_VERSION,
                       "Ben Pfaff", "John Darrington", NULL_SENTINEL);
          exit (EXIT_SUCCESS);

        case 'h':
          usage ();
          exit (EXIT_SUCCESS);

        default:
          goto error;
        }
    }

  if (optind + 2 != argc)
    error (1, 0, _("exactly two non-option arguments are required; "
                   "use --help for help"));

  input_filename = argv[optind];
  output_filename = argv[optind + 1];
  input_fh = fh_create_file (NULL, input_filename, NULL, fh_default_properties ());

  if (output_format == NULL)
    {
      const char *dot = strrchr (output_filename, '.');
      if (dot == NULL)
        error (1, 0, _("%s: cannot guess output format (use -O option)"),
               output_filename);

      output_format = dot + 1;
    }

  output_fh = fh_create_file (NULL, output_filename, NULL, fh_default_properties ());
  if (encrypted_file_open (&enc, input_fh) > 0)
    {
      if (decrypt_file (enc, input_fh, output_fh, password,
                         ds_cstr (&alphabet), length, password_list))
        goto exit;
      else
	goto error;
    }


  reader = any_reader_open_and_decode (input_fh, encoding, &dict, NULL);
  if (reader == NULL)
    goto error;

  if (keep)
    {
      struct variable **keep_vars;
      size_t n_keep_vars;
      if (!parse_variables_option (keep, dict, &keep_vars, &n_keep_vars))
        goto error;
      dict_reorder_vars (dict, keep_vars, n_keep_vars);
      dict_delete_consecutive_vars (dict, n_keep_vars,
                                    dict_get_var_cnt (dict) - n_keep_vars);
      free (keep_vars);
    }

  if (drop)
    {
      struct variable **drop_vars;
      size_t n_drop_vars;
      if (!parse_variables_option (drop, dict, &drop_vars, &n_drop_vars))
        goto error;
      dict_delete_vars (dict, drop_vars, n_drop_vars);
      free (drop_vars);
    }

  if (!strcmp (output_format, "csv") || !strcmp (output_format, "txt"))
    {
      if (!csv_opts.delimiter)
        csv_opts.delimiter = csv_opts.decimal == '.' ? ',' : ';';
      writer = csv_writer_open (output_fh, dict, &csv_opts);
    }
  else if (!strcmp (output_format, "sav") || !strcmp (output_format, "sys"))
    {
      struct sfm_write_options options;

      options = sfm_writer_default_options ();
      writer = sfm_open_writer (output_fh, dict, options);
    }
  else if (!strcmp (output_format, "por"))
    {
      struct pfm_write_options options;

      options = pfm_writer_default_options ();
      writer = pfm_open_writer (output_fh, dict, options);
    }
  else
    {
      error (1, 0, _("%s: unknown output format (use -O option)"),
             output_filename);
      NOT_REACHED ();
    }
  if (!writer)
    error (1, 0, _("%s: error opening output file"), output_filename);

  for (i = 0; i < max_cases; i++)
    {
      struct ccase *c;

      c = casereader_read (reader);
      if (c == NULL)
        break;

      casewriter_write (writer, c);
    }

  if (!casereader_destroy (reader))
    error (1, 0, _("%s: error reading input file"), input_filename);
  if (!casewriter_destroy (writer))
    error (1, 0, _("%s: error writing output file"), output_filename);

exit:
  ds_destroy (&alphabet);
  dict_unref (dict);
  fh_unref (output_fh);
  fh_unref (input_fh);
  fh_done ();
  i18n_done ();

  return 0;

error:
  casereader_destroy (reader);
  ds_destroy (&alphabet);
  dict_unref (dict);
  fh_unref (output_fh);
  fh_unref (input_fh);
  fh_done ();
  i18n_done ();

  return 1;
}

static bool
decrypt_file (struct encrypted_file *enc,
	      const struct file_handle *ifh,
	      const struct file_handle *ofh,
              const char *password,
              const char *alphabet,
              int max_length,
              const char *password_list)
{
  FILE *out;
  int err;
  const char *input_filename = fh_get_file_name (ifh);
  const char *output_filename = fh_get_file_name (ofh);

  if (password_list)
    {
      FILE *password_file;
      if (!strcmp (password_list, "-"))
        password_file = stdin;
      else
        {
          password_file = fopen (password_list, "r");
          if (!password_file)
            error (1, errno, _("%s: error opening password file"),
                   password_list);
        }

      struct string pw = DS_EMPTY_INITIALIZER;
      unsigned int target = 100000;
      for (unsigned int i = 0; ; i++)
        {
          ds_clear (&pw);
          if (!ds_read_line (&pw, password_file, SIZE_MAX))
            {
              if (isatty (STDOUT_FILENO))
                {
                  putchar ('\r');
                  fflush (stdout);
                }
              error (1, 0, _("\n%s: password not in file"), password_list);
            }
          ds_chomp_byte (&pw, '\n');

          if (i >= target)
            {
              target += 100000;
              if (isatty (STDOUT_FILENO))
                {
                  printf ("\r%u", i);
                  fflush (stdout);
                }
            }

          if (encrypted_file_unlock__ (enc, ds_cstr (&pw)))
            {
              printf ("\npassword is: \"%s\"\n", ds_cstr (&pw));
              password = ds_cstr (&pw);
              break;
            }
        }
    }
  else if (alphabet[0] && max_length)
    {
      size_t alphabet_size = strlen (alphabet);
      char *pw = xmalloc (max_length + 1);
      int *indexes = xzalloc (max_length * sizeof *indexes);

      for (int len = password ? strlen (password) : 0;
           len <= max_length; len++)
        {
          if (password && len == strlen (password))
            {
              for (int i = 0; i < len; i++)
                {
                  const char *p = strchr (alphabet, password[i]);
                  if (!p)
                    error (1, 0, _("%s: '%c' is not in alphabet"),
                           password, password[i]);
                  indexes[i] = p - alphabet;
                  pw[i] = *p;
                }
            }
          else
            {
              memset (indexes, 0, len * sizeof *indexes);
              for (int i = 0; i < len; i++)
                pw[i] = alphabet[0];
            }
          pw[len] = '\0';

          unsigned int target = 0;
          for (unsigned int j = 0; ; j++)
            {
              if (j >= target)
                {
                  target += 100000;
                  if (isatty (STDOUT_FILENO))
                    {
                      printf ("\rlength %d: %s", len, pw);
                      fflush (stdout);
                    }
                }
              if (encrypted_file_unlock__ (enc, pw))
                {
                  printf ("\npassword is: \"%s\"\n", pw);
                  password = pw;
                  goto success;
                }

              int i;
              for (i = 0; i < len; i++)
                if (++indexes[i] < alphabet_size)
                  {
                    pw[i] = alphabet[indexes[i]];
                    break;
                  }
                else
                  {
                    indexes[i] = 0;
                    pw[i] = alphabet[indexes[i]];
                  }
              if (i == len)
                break;
            }
        }
      free (indexes);
      free (pw);

    success:;
    }
  else
    {
      if (password == NULL)
        {
          password = getpass ("password: ");
          if (password == NULL)
            return false;
        }

      if (!encrypted_file_unlock (enc, password))
        error (1, 0, _("sorry, wrong password"));
    }

  out = fn_open (ofh, "wb");
  if (out == NULL)
    error (1, errno, ("%s: error opening output file"), output_filename);

  for (;;)
    {
      uint8_t buffer[1024];
      size_t n;

      n = encrypted_file_read (enc, buffer, sizeof buffer);
      if (n == 0)
        break;

      if (fwrite (buffer, 1, n, out) != n)
        error (1, errno, ("%s: write error"), output_filename);
    }

  err = encrypted_file_close (enc);
  if (err)
    error (1, err, ("%s: read error"), input_filename);

  if (fflush (out) == EOF)
    error (1, errno, ("%s: write error"), output_filename);
  fn_close (ofh, out);

  return true;
}

static void
usage (void)
{
  printf ("\
%s, a utility for converting SPSS data files to other formats.\n\
Usage: %s [OPTION]... INPUT OUTPUT\n\
where INPUT is an SPSS data file or encrypted syntax file\n\
  and OUTPUT is the name of the desired output file.\n\
\n\
The desired format of OUTPUT is by default inferred from its extension:\n\
  csv txt             comma-separated value\n\
  sav sys             SPSS system file\n\
  por                 SPSS portable file\n\
  sps                 SPSS syntax file (encrypted syntax input files only)\n\
\n\
General options:\n\
  -O, --output-format=FORMAT  set specific output format, where FORMAT\n\
                      is one of the extensions listed above\n\
  -e, --encoding=CHARSET  override encoding of input data file\n\
  -c MAXCASES         limit number of cases to copy (default is all cases)\n\
  -k, --keep=VAR...   include only the given variables in output\n\
  -d, --drop=VAR...   drop the given variables from output\n\
CSV output options:\n\
  --recode            convert user-missing values to system-missing\n\
  --no-var-names      do not include variable names as first row\n\
  --labels            write value labels to output\n\
  --print-formats     honor variables' print formats\n\
  --decimal=CHAR      use CHAR as the decimal point (default: .)\n\
  --delimiter=CHAR    use CHAR to separate fields (default: ,)\n\
  --qualifier=CHAR    use CHAR to quote the delimiter (default: \")\n\
Password options (for used with encrypted files):\n\
  -p PASSWORD         individual password\n\
  -a ALPHABET         with -l, alphabet of passwords to try\n\
  -l MAX-LENGTH       with -a, maximum number of characters to try\n\
  --password-list=FILE  try all of the passwords in FILE (one per line)\n\
Other options:\n\
  --help              display this help and exit\n\
  --version           output version information and exit\n",
          program_name, program_name);
}
